### ISSUE TITLE
Add option to use [gpi]_printf to log through the simulator

### ISSUE THREAD

**Issue by chiggs** [cocotb/cocotb#23]
I want cocotb output to go to my console, not the VCS window

**Comment by GordonMcGregor** [comment]
Been trying to integrate the 1.0 release (congrats btw) into our UVM, VCS, SVerilog based flow.  Was easy to get up and running, but I was quickly reminded of the different ideas of stdout between VCS and the python interpreter.  As we capture output using the VCS log switch, it would be great to have an easy way to unify the sys.stdout filehandle in python that the Logger is using, with the stdout that VCS is using.

Was any progress made on providing a way to switch these?

**Comment by chiggs** [comment]
Sorry, nothing has been done on this.

Shouldn't be very difficult to knock something up though, if I get a spare half hour I'll take a look.

**Comment by chiggs** [comment]
@GordonMcGregor can you see whether edcaf84 does what you want?

If so I'll plumb in VHPI and FLI (currently only done VPI) and make this the default log mechanism.

**Comment by chiggs** [comment]
Hmmmm, travis didn't like edcaf84 due to a newer GCC with embiggened warnings.  Works fine on CentOS 6.5 though.  Possibly need to be less lazy and use variadic function after all....

**Comment by GordonMcGregor** [comment]
Yes, that seems to work. I'd partially implemented things so handled replacing the StreamHandler slightly differently, passing it a 'simulator stdout' replacement file object.  Not sure which is more inline with that standard approaches using logging. 

But the end result is the same, all the log messages from systemverilog and CocoTb end up in the one combined run log, which is a big improvement.

```python
class VpiStdout(object):
    def write(self, str):
        simulator.sim_printf(str)
    def flush(self):
        pass

class SimBaseLog(logging.getLoggerClass()):
    def __init__(self, name):
        hdlr = logging.StreamHandler(VpiStdout())
```

**Comment by ktbarrett** [comment]
Seems like this requires simulator-specific support to work through `stdio`, which can be passed via `SIM_ARGS` and friends. And if you wanted to log to another location, say a file, that is now supported by adding another handler to the cocotb logger instance.

**Comment by eric-wieser** [comment]
I'd be inclined to reopen this - the distinction between the simulator console and the terminal stdout is annoying, and it would be great if we could at least document how to set up such a logging handler.

**Comment by marlonjames** [comment]
I think it's worth considering the feature of using the `vpi_printf`/`vhpi_printf`/`mti_PrintMessage` functions for logging. See https://github.com/cocotb/cocotb/issues/2015#issuecomment-1027304668
